### DISCUSSION TITLE
Improved data flow and lifetimes

### DISCUSSION THREAD

**Discussion by tony-p** [BehaviorTree/BehaviorTree.CPP#620] [Ideas]
I am currently looking into BT.CPP for a project but am hesitant due to the blackboard model for sharing data. Having a general blackboard for sharing data is like throwing all data as globals into the application. This is generally bad because:
* Risk of unintended side-effects - A second action could modify the leading to a third action acting on the wrong result
* Out of date data - An action acts on data that was generated too long ago that it is no longer valid
* Lack of observability - With no direct link between producers and consumers, can be hard to follow how data is used.
* Composability/reusablility - No guarantees that globals don't overlap

To solve at least some of these issues I would propose that data is piped directly from output ports to input ports. This would show clearly the link from producers to consumers, would also make clear exactly what version of data is used at an action (no in between modifiers).
This in itself doesn't solve the out of date data issue, however for simple downstream use cases, resetting triggers could be achieved (would need to think about more).

I don't necessarily think this will remove the need for the blackboard entirely (there may be a need for a few application wide variables) however the blackboard approach doesn't scale to larger projects and teams.

**Comment by facontidavide** [comment]
Hi,

BT.CPP already improves over the model of the "classical" blackboard providing the concept of Input/Output ports.

This is the reason why people MUST not use the blackboard directly.

There are certainly way to imorive it, but, to be fair:

- **unintended side-effect**: that would be true with other models such as publish-subscribe.
- **Out of date data**: this is a decision that was taken, and people may or may not consider it a feature.
- **Observability**: we can certainly add the ability to log changes and the connection between producer and consumer is actually known at deployment time.
- **Composability/reusablility** not entireely true, since BT.CPP does have blackboard "scoping" at the level of the subtree.

> To solve at least some of these issues I would propose that data is piped directly from output ports to input ports.

I can certainly implement that, but I am not sure how that will improve the cocerns mentioned above.

The only advantage of  a publish-subscribe model is that the subscriber can access all the publications using a FIFO queue, instead or reading the latest value, only.

A preliminary step would be to create an abstract class for the "backend" of Input/Output ports, where the blackboard is only one among multiple implementations.

If anyone want to invest time (or money) into investigation this, I will be happy to discuss it further.

**Reply by tony-p** [reply]
> I can certainly implement that, but I am not sure how that will improve the cocerns mentioned above.

I wasn't really talking about a pub/sub model, but more direct connection of the ports (essentially more a socket model). A naive implementation example could be achieved via the blackboard by generating mangled names (if using a UI) so that only ever 1 producer can write to a particular variable. However this naive implementation would limit the ability to accept an input from multiple sources depending on execution path. This would prevent unintended side effects from occurring (IE you know the source of your data as only 1 node could have ever written it).

If you visualize it from the UI point of view, instead of only connecting nodes by the relationship (tree structure) you also connect lines from the outputs to inputs.

> Composability/reusablility not entireely true, since BT.CPP does have blackboard "scoping" at the level of the subtree.

Indeed I only spotted this functionality after writing this comment. This does help to make the scope smaller and hence comprehensible to limit the chance of side effects occurring as well, although still not exactly explicit.

For now I will have a play and see how it is to use
